### PR TITLE
Merge ML_API Branch from Sprint 1

### LINKED ISSUE #6: ML - Training Script - Linear Regression
## User Story
As a user, I would like to train a linear regression model. I want to be able to go to my dataset, select a numerical label, and train a Linear Regression model.

### Persona
Existing users looking to train a model.

### Feature
Be able to run a training job through our app, specifically for a linear regression model.

### Business Value
Users can train a new model type from our app for numerical datasets and regression problems.

## Tasks

1. Script takes in input hyperparameters
2. Script loads the dataset
3. Script initializes a model & trains on the dataset
4. Script saves/uploads the trained model

## Acceptance Criteria
- Will NOT implement integration with our Rails app. Simply ML scripts to train a given model.

## Acceptance Tests

- Users can schedule a Linear Regression model training
- Training runs successfully
- Final model is saved in an appropriate location

### LINKED ISSUE #7: ML - Training Script - Logistic Regression
## User Story
As a user, I would like to train a logistic regression model. I want to be able to go to my dataset, select a categorical label, and train a Logistic Regression model.

### Persona 
Existing users looking to train a model.

### Feature
Be able to run a training job through our app, specifically for a logistic regression model.

### Business Value
Users can train a new model type from our app for datasets with categorical labels.

## Tasks

1. Script takes in input hyperparameters
2. Script loads the dataset
3. Script initializes a logistic regression model & trains on the dataset
4. Script saves/uploads the trained model

## Acceptance Criteria
- Will NOT implement integration with our Rails app. Simply ML scripts to train a given model.

## Acceptance Tests

- Users can schedule a Logistic Regression model training
- Training runs successfully
- Final model is saved in an appropriate location

### LINKED ISSUE #8: ML - Training Script - Decision Trees
## User Story
As a user, I would like to train a decision tree model. I want to be able to go to my dataset, select a label, choose amongst some pair of hyperparameters, and train a Decision Tree model.

### Persona 
Existing users looking to train a model.

### Feature
Be able to run a training job through our app, specifically for a decision tree model.

### Business Value
Users can train a new model type from our app for datasets with categorical labels.

## Tasks

1. Script takes in input hyperparameters
2. Script loads the dataset
3. Script initializes a decision tree model & trains on the dataset
4. Script saves/uploads the trained model

## Acceptance Criteria
- Will NOT implement integration with our Rails app. Simply ML scripts to train a given model.

## Acceptance Tests

- Users can schedule a Decision Tree model training
- Training runs successfully
- Final model is saved in an appropriate location

### PR DESCRIPTION
**Pull request by congbrian**
This merge covers issues #6 #7 #8 and merges branch [ML_API in github/MLite](https://github.com/COSC481W-2024Fall/MLite).

**Features**
1. Utilizes iris_setosa dataset as dummy data via sklearn API. Iris dataset is a widely used machine learning dataset often used for starter projects first recorded by Fisher, 1936, and contains a variety of measurements of irises alongside their species classification.
2. Converts iris_setosa dummy data from sklearn np array to pandas.dataFrame to ensure ease of use and data analysis. This is critical as all dataset from now will use pandas as a medium for cleaning and preprocessing, simply due to ease of usage; pandas data is more readable, easier to access, and has more in depth tabular functions than a raw numpy array.
3. Utilizes sklearn libraries to call linear regression as per #6 , in addition utilizes a one-hot encoding technique from pandas library (pandas.get_dummies) to encode iris species column into one hot data, which allows it to be used as a predictor in linear regression.
4. Utilizes sklearn library to call logistic regression as per #7 ; logistic regression is inherently categorical and therefore fits iris dataset well.
5. Utilizes sklearn library to call decision trees as per #8 ; decision trees are technically more suited to categorical data and therefore fits iris dataset well without need for significant adaptations.
6. Automatically splits columns to segregate features and labels; this is a vertical split -- it splits the column(s) to be predicted by the algorithm away from the columns being used to predict and prevents data leakage before feeding them into the algorithms.
7. Automatically splits columns to segregate train/test datasets; this is a horizontal split, which, if done correctly, will create an independent test set that can be used by the created algorithm to check its final accuracy after being trained on data that it has never seen before, and therefore how generalizable the model outcome is.
8. Does all of this within an API class called MLAPI that returns a model, checks for GPU acceleration capability, and can associate a dataset to itself.

**Code**

- Adds one notebook which is used to prototype code but cannot interface directly with a pipeline/containerization architecture.
- Uses a notebook call that allows direct conversion to generate a python script that can interface with data pipelines.

**Testing**
- Each class function includes its own testing and validation as per standard machine learning practices -- it will output the final test set accuracy when running after finishing all epochs.
- Robustly checks for CUDA acceleration utilizing a try-catch loop to account for exceptions.

**What to look out for**
- Several TODOs on this particular class that will need to be handled.
- Worth noting that this class will likely serve as the primary backend class for the entire project, and will likely grow to be fairly long if a wide variety of models are implemented.
- Keep in mind that the model will hold a dataset and keeps/returns model weights associated with the instance of each class.
- Sklearn does not inherently utilize GPU acceleration and typically runs on CPU due to how lightweight many of its algorithms on, but has some GPU acceleration features. 
- We will definitely need to pipe GPU access through the container. Am familiar with doing that process in Docker, but unsure of how resources for that will look on the cloud, nor if we will even be granted access to use them during this project, in which case our training capability will be limited to fairly small models.

### CODE REVIEW

**Review by Mohammad4844 — APPROVED** [review]
Everything looks good. Tests run as expeted. Good to merge.
